### PR TITLE
Add some useful Transaction methods to handle nLockTime

### PR DESCRIPTION
**Pull request by maraoz**
This PR also:
- fixes 0 fee handling
- fixes change address logic
- adds Transaction#getChangeOutput

### CODE REVIEW

**Comment by maraoz** [comment]
updated

**Comment by eordano** [comment]
We should add a method to check if the transaction is valid (based only on the nLockTime and only if it's a timestamp). We *could* add an estimated (height -> timestamp) conversion based on historical data and expected difficulty growth?

**Comment by eordano** [comment]
Coverage decreased (-0.01%) to 97.1%

**Comment by maraoz** [comment]
fixed coverage, ready for review :)

File: lib/transaction/transaction.js [hunk 1]
```diff
@@ -269,14 +257,84 @@ Transaction.prototype.fromObject = function(transaction) {
...
+};
+
+Transaction.prototype._checkConsistency = function() {
+  if (!_.isUndefined(this._changeIndex)) {
+    $.checkState(this._changeScript);
+    $.checkState(this.outputs[this._changeIndex]);
+    $.checkState(this.outputs[this._changeIndex].script.toString() ===
+      this._changeScript.toString());
+  }
+  // TODO: add other checks
```
Comment (braydonf) [inline lib/transaction/transaction.js]: What other checks? An issue might be better, and more descriptive.

Comment (maraoz) [inline lib/transaction/transaction.js]: yes, was planning to open an issue, thanks for the reminder :)

File: test/transaction/transaction.js [hunk 2]
```diff
@@ -120,6 +120,8 @@ describe('Transaction', function() {
     script: Script.buildPublicKeyHashOut(fromAddress).toString(),
     satoshis: 100000
   };
+  var weirdUtxoWith100000Satoshis = JSON.parse(JSON.stringify(simpleUtxoWith100000Satoshis));
+  weirdUtxoWith100000Satoshis.script = new Script().add('OP_TRUE');
```
Comment (braydonf) [inline test/transaction/transaction.js]: `var anyoneCanSpendUtxo = `?

File: test/transaction/transaction.js [hunk 3]
```diff
@@ -407,6 +436,66 @@ describe('Transaction', function() {
...
+        return new Transaction().lockUntilDate(499999999);
+      }).to.throw(errors.Transaction.LockTimeTooEarly);
+    });
+    it('fails if the block height is negative', function() {
+      expect(function() {
+        return new Transaction().lockUntilBlockHeight(-1);
+      }).to.throw(errors.Transaction.NLockTimeOutOfRange);
+    });
+  });
+  it('handles weird output', function() {
```
Comment (braydonf) [inline test/transaction/transaction.js]: Same as above, "handles an anyone-can-spend output" could be more descriptive.

File: lib/transaction/transaction.js [hunk 4]
```diff
@@ -269,14 +257,84 @@ Transaction.prototype.fromObject = function(transaction) {
...
+};
+
+/**
+ * Sets nLockTime so that transaction is not valid until the desired date(a
+ * timestamp in seconds since UNIX epoch is also accepted)
+ *
+ * @param {Date | Number} time
+ * @return {Transaction} this
```
Comment (braydonf) [inline lib/transaction/transaction.js]: These may need to be `@returns`

**Comment by braydonf** [comment]
LGTM

**Comment by eordano** [comment]
LGTM

File: lib/transaction/transaction.js (hunk 4 above)
Comment (maraoz) [inline lib/transaction/transaction.js]: apparently they're synonyms http://usejsdoc.org/tags-returns.html

**Comment by maraoz** [comment]
updated with @braydonf comments

**Pull request merged**